Audio Driver using OpenAL

HilbertCurve opened the pull request:
In this PR, I coded up an audio driver using OpenAL. The audio driver is split into four main parts:

1. Audio Buffer: Using buffer utils, I pull the audio data from the assets folder and store it in an array. This data is then to be used by an audio source.
2. Audio Source: This class is a Component; it's a point in 3D space (or 2D for our case) where noise can come from. It also has this neat property where, if you set the `AL_VELOCITY` property using `alSourcef()`, OpenAL can apply the Doppler Effect to the outputted noise, in case we want to use that. We would attach this to a gameObject first to enable properties like it's position (this is easy to implement, I just haven't gotten around to it yet.)
3. Audio Listener: This class is also a Component. It is what acts like an "ear" of some sorts. All noise calculations are done with the listener in mind (whether to play the noise on the right or left speaker, relative velocity, etc.) and there can only be one listener; that's why its a singleton.
4. Audio Master: Finally, this is where all the `AudioSource`'s are. At the end of runtime, some amount of cleanup is required before closing the program (sure, the library does it itself, but I'd prefer to do it manualy.)

I tried to document most classes; that may need a little work, though. I hope this is a good start to the audio driver.

Juyas commented on src/main/test/audio/AudioTest.java:
```diff
@@ -1,26 +1,33 @@
 
+        try {
+            // simulate things happening in game
+            Thread.sleep(10000);
```
I suggest to never do this in a test - a test is never supposed to be running longer than the code takes - artifical waiting time slows down the compilation process, which will get slower over time anyways.

HilbertCurve replied:
You're right, I'll probably make another scene as a test instead of this when I can get to it.

Juyas replied:
yes, manual tests are preferable for OpenAL as well as everything related with OpenGL and so on

Juyas commented on src/main/java/audio/AudioMaster.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public static AudioMaster get() {
+        if (instance == null) {
```
I suggest making the inner part of instance==null synchronized - to prevent two get calls possibly running parallel

HilbertCurve commented:
I synched up the singleton getters, removed `AudioTest`, fixed some issues with `AudioListener` and `AudioSource`, and made a dedicated scene for testing the driver.

HilbertCurve commented:
Added a dedicated scene for testing the driver.

HilbertCurve commented:
Yeah, I don't think an audio batch makes sense: you're gonna call `alPlaySource()` for each sound you play no matter what.

athaun commented on src/main/java/audio/AudioSource.java:
```diff
@@ -76,7 +77,6 @@ public boolean isLooping() {
      * this buffer.
      */
     public void play(int index) {
-        System.out.println("hi");
```
my favorite way to debug :D

HilbertCurve commented:
Audio streaming basically means your sound card isn't overloaded with 2.3 megabytes of data whenever you play a song; instead 8 chunks of data are queued and then refreshed once processed

athaun commented:
Nice, I will try to take a look soon.

athaun reviewed (commented):
I have looked through the code and done some preliminary testing, so far things are looking good.

athaun commented:
This is a good start for sure, and I am ready to merge. In the future, if you want to add more contributions, better error handling (not crashing when it can't find a file), wider filetype support, and spatial audio would be awesome.
